Making _MeshEnv subclassing thread local

wz337 opened the pull request:
With _mesh_resources being global var, when thread pg based testing is used (aka spawn_threads_and_init_comms()), the last rank with the same key would overwrite the formers. This isn't an issue in regular process-based runtime as logically each key is unique.

Example failure: https://github.com/pytorch/pytorch/actions/runs/8779134353/job/24087295785
```
RuntimeError: Could not resolve the process group registered under the name 8
or
Throwing assert not none error 
```




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @tianyu-l @wconstab @yf225 @chauhang @d4l3k

awgu commented:
Does this mean that MTPG tests have just been flaky before this? I wonder do you know how many tests were affected?

wz337 commented:
> Does this mean that MTPG tests have just been flaky before this? I wonder do you know how many tests were affected?

So far it seems there are two issues for MTPG tests: 
1. Shawn saw some race condition in threaded pg in some APF tests, which is threaded pg related and not DeviceMesh related.  
2. _mesh_resources being a global var seems be problematic in threaded backend, as the next rank would overwrite the previous, which hopefully this PR can resolve. 

My feeling is that this could cause all threaded pg + DeviceMesh test being flaky. Afaik, we don't have too many tests using these two together yet. From what I remember, there are:
- test_dtensor_ops(https://github.com/pytorch/pytorch/blob/main/test/distributed/_tensor/test_dtensor_ops.py), which could explain why we periodically see some test fail. https://github.com/pytorch/pytorch/issues/123294
- the FSDP2 tests with FSDPTestMultiThread. Hence, we are seeing some of the tests being flaky. https://github.com/pytorch/pytorch/issues/121233 

Running multi process tests seems more reliable, not sure how much we care about CI test performance, but hopefully this change could resolve the issue, so we can get more accurate test signal.

wz337 committed:
Making _MeshEnv subclassing thread local

Summary:
With _mesh_resources being global var, when thread pg based testing is used (aka spawn_threads_and_init_comms()), the last rank with the same key would overwrite the formers. This isn't an issue in regular process-based runtime as logically each key is unique.

Example failure: https://github.com/pytorch/pytorch/actions/runs/8779134353/job/24087295785

Test Plan: CI

Reviewed By: wanchaol

Differential Revision: D56327061
